### PR TITLE
Okta-594842 : Custom password expiry

### PR DESCRIPTION
**Pull request by leemchale-okta**
## Description:

This PR implements the Custom password expiry remediation transformer aswell as the Custom password expiry warning transformer in SIW gen 3.

## PR Checklist

- [x] Have you verified the basic functionality for this change?
- [x] Did you add tests, as appropriate, following our [Automated Test guidelines](https://oktawiki.atlassian.net/wiki/spaces/eng/pages/2676497890/Automated+Testing+in+the+Signin+Widget)?
- [ ] Did you follow our [Security Best Practices](https://oktawiki.atlassian.net/wiki/display/eng/Security+Best+practices)?
- [ ] Did you verify the change by running [downstream monolith artifact](https://oktawiki.atlassian.net/wiki/spaces/eng/pages/102897979/Sign-in+Widget+Development#Sign-inWidgetDevelopment-Instructionstocreateandrunthedownstreamartifact(d16t))? (Provide link to build below)
- [x] Does this PR include noticeable changes to the UI? (If yes, attach screenshots/video below)

### Issue:

- [OKTA-594842](https://oktainc.atlassian.net/browse/OKTA-594842)
- [OKTA-594840](https://oktainc.atlassian.net/browse/OKTA-594840)

### Screenshot/Video:

Custom password expiry:
Screenshot 2023-04-10 at 11 42 22 AM

Custom password expiry warning:
Screenshot 2023-04-10 at 11 41 52 AM

### CODE REVIEW

File: playground/mocks/data/idp/idx/authenticator-expired-custom-password.json [hunk 1]
```diff
@@ -38,7 +38,7 @@
           }
         ],
         "accepts": "application/json; okta-version=1.0.0",
-        "customExpiredPasswordName": "expiredpasswordOIE",
+        "customExpiredPasswordName": "Password reset website name",
```
Comment (leemchale-okta) [inline playground/mocks/data/idp/idx/authenticator-expired-custom-password.json:41]: changed mock value here to match `customExpiredPasswordName` in Custom password expiry warning mock file to make it more consistent.  Also so only one spec test object method is needed to select the button

File: src/v3/src/transformer/password/transformReEnrollCustomPasswordExpiryWarning.ts [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+    if(typeof skipStep !== 'undefined') {
+      uischema.elements.push({
+        type: 'Link',
+        contentType: 'footer',
+        options: {
+          label: loc('password.expiring.later', 'login'),
+          isActionStep: true,
+          step: skipStep?.name,
+        },
+      } as LinkElement)
```
Comment (glenfannin-okta) [inline src/v3/src/transformer/password/transformReEnrollCustomPasswordExpiryWarning.ts:99]:
```suggestion
      } as LinkElement);
```

File: src/v3/src/transformer/password/transformReEnrollCustomPasswordExpiryWarning.ts [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+
+  const getPasswordExpiryTitle = (daysToExpiry = -1): string => {
+    if (daysToExpiry > 0) {
+      return loc('password.expiring.title', 'login', [daysToExpiry]);
+    } else if (daysToExpiry === 0) {
+      return loc('password.expiring.today', 'login');
+    } else {
+      return loc('password.expiring.soon', 'login');
+    }
+  };
```
Comment (glenfannin-okta) [inline src/v3/src/transformer/password/transformReEnrollCustomPasswordExpiryWarning.ts:32]: I think what you could do is extract [this](https://github.com/okta/okta-signin-widget/blob/0.5/src/v3/src/transformer/password/transformExpiredPasswordWarningAuthenticator.ts#L26-L36) function into a utils file and use it here and in the original location?

File: src/v3/src/transformer/password/transformReEnrollCustomPasswordExpiryWarning.test.ts [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+      .toBe('password.expired.custom.submit');
+  });
+
+  // Days to expiry, correct title
+  [
+    [0, 'password.expiring.today'],
+    [4, 'password.expiring.title'],
+    [null, 'password.expiring.soon'],
+  ].forEach(([ daysToExpiry, formTitle ]) => {
+    it('should add correct title with this many days to expiry: ' + daysToExpiry, () => {
```
Comment (glenfannin-okta) [inline src/v3/src/transformer/password/transformReEnrollCustomPasswordExpiryWarning.test.ts:67]: There is a built-in way to do this type of test, see: https://jestjs.io/docs/api#testeachtablename-fn-timeout, you can see some examples here: https://dev.to/bgord/simplify-repetitive-jest-test-cases-with-test-each-310m

File: src/v3/src/transformer/password/transformReEnrollCustomPasswordExpiryWarning.ts [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+    const subtitle = brandName
+      ? loc('password.expiring.subtitle.specific', 'login', [brandName])
+      : loc('password.expiring.subtitle.generic', 'login');
+    const subtitleElement: DescriptionElement = {
+      type: 'Description',
+      contentType: 'subtitle',
+      options: {
+        content: subtitle + ' ' + loc('password.expired.custom.subtitle', 'login'),
+      },
+    };
```
[the remark this answers is not in the document]
Comment (leemchale-okta) [reply src/v3/src/transformer/password/transformReEnrollCustomPasswordExpiryWarning.ts:69]: ah awesome thank you!

File: src/v3/src/transformer/password/transformReEnrollCustomPasswordExpiryWarning.test.ts (hunk 4 above)
Comment (leemchale-okta) [reply src/v3/src/transformer/password/transformReEnrollCustomPasswordExpiryWarning.test.ts]: oh awesome thank you

**Review by glenfannin-okta — APPROVED** [review]

File: packages/@okta/i18n/src/properties/login.properties [hunk 6]
```diff
@@ -710,6 +710,10 @@ password.expiring.subtitle.specific = When password expires you will be locked o
 password.expired.custom.submit = Go to {0}
 password.expired.custom.subtitle = This password is set on another website. Click the button below to go there and set a new password.
 
+## Custom Password expiring soon.
+password.expiring.soon.subtitle.generic = When password expires you will be locked out of your account. This password is set on another website. Click the button below to go there and set a new password.
+password.expiring.soon.subtitle.specific = When password expires you will be locked out of your {0} account. This password is set on another website. Click the button below to go there and set a new password.
```
Comment (shawyu-okta) [inline packages/@okta/i18n/src/properties/login.properties:715]: Are these translations coming from `master`?

File: test/testcafe/framework/page-objects/EnrollCustomPasswordPageObject.js [hunk 7]
```diff
@@ -1,17 +1,24 @@
...
-    await this.form.clickElement(REDIRECT_BUTTON);
+    await this.t.click(this.getRedirectButton());
+  }
+
+  getRedirectButton() {
+    if(userVariables.v3) {
+      return this.form.getButton('Go to Password reset website name');
+    }
+
+    return this.form.getLink('Go to Password reset website name');
```
[the remark this answers is not in the document]
Comment (shawyu-okta) [reply test/testcafe/framework/page-objects/EnrollCustomPasswordPageObject.js:18]: > I assume they just format the link to look like a button in v2?

yeah exactly. there's a lot of cases in g2 where there is an anchor styled like a button. i'd like us to address the a11y of how we use buttons as links in g3 (the inverse issue) but i think that's a separate conversation from the work here

File: packages/@okta/i18n/src/properties/login.properties (hunk 6 above)
Comment (leemchale-okta) [reply packages/@okta/i18n/src/properties/login.properties:715]: yup they are coming from this change pr from master: https://github.com/okta/okta-signin-widget/pull/3171/files

File: test/testcafe/spec-en-leaks/EnglishLeaks_spec.js [hunk 8]
```diff
@@ -31,8 +31,6 @@ const ignoredMocks = [
 ];
 
 const v3IgnoredMocks = [
-  'authenticator-expiry-warning-custom-password.json', // TODO: Not yet implemented OKTA-594840
```
Comment (nikhilvenkatraman-okta) [inline test/testcafe/spec-en-leaks/EnglishLeaks_spec.js:34]: 👍

**Review by nikhilvenkatraman-okta — COMMENTED** [review]
i18n changes look good

**Pull request merged**